docs_src/design/adr/core/0026-Core-Data-Cache.md

```diff
@@ -0,0 +1,85 @@
+# Core Data Cache
+
+## Status
+
+**Draft**
+
+proposed: 4/4/22
+
+## Context
+There is a desire to have core data serve more as a data cache - that is, keeping only the "latest" readings.  There is also a desire to allow for queries of core data to serve up the "latest" readings.  This ADR defines new operations and APIs to satisfy these needs.
+
+### Queries to return "Latest"
+
+Core data will add the following new operations (and externally visible APIs) to provide the means to get the `latest` readings.
+
+- Query for the latest N readings by device name
+    - API: *GET* /reading/device/name/{name}/latest/{N}
+    - returns the last N readings for the specified device
+    - N must be > 0
+- Query for the latest N readings by device name and device resource
+    - API: *GET* /reading/device/name/{name}/resourceName/{resourceName}/latest/{N}
+    - returns the latest N readings for the specified device and specified device resource
+    - N must be > 0
+
+### Delete operations (to keep the number of readings to the "latest" or a database size limit)
+
+Core data will add the following new delete operations.  Note that only one will be exposed externally via API.  The other will be used internally by the service (when configured to use it).
+
+- Delete all but N readings (aka - the **Eaton Delete**; clean readings but preserve the latest readings)
+    - API:  *DELETE* /reading/keep/{N}
+    - if there are fewer than (or equal to) N readings in core data, then all readings would be kept and no readings would be removed
+    - if there are more than N readings in core data, then delete the oldest readings, but keep the youngest N readings
+    - N must be > 0 (N = 1 means keep only the latest reading)
+    - this delete option will be triggered via API either:
+        - manually by the adopter
+        - or by the scheduling service (the adopter would set up a new IntervalAction with the scheduling service)
+
+- Delete readings to keep the database at a certain size (aka - the **Tony Delete**; clean readings to preserve the number of readings in the database)
```

Comment:
The unit of deletion should be `Event`, not ` Reading`.
If one `Event` contains 1000 `Readings`, how could we delete some of the `Readings`?
That's why we don't have the Delete Reading API.

Reply:
It is difficult I think to make sure that we can meet the "delete all but N readings" when events are used. It may require even more processing. I think you cover this processing challenge below.

Reply:
yeah, I agree with that.  if the user only need readings, we can allow `Events` with missing `Readings`

Reply:
Agreed. Reducing processor utilization for "pruning" would be a good thing. I believe we included the keep N because it seemed "easy" at the time. Is it causing more pain than it is worth? Is it better/simpler/more efficient to just keep the latest?